R/method-correlation.R

```diff
@@ -0,0 +1,100 @@
+#' @param verbose boolean indicating if timed logging is desired
+#' @return ComputeResult object
+#' @import veupathUtils
+#' @import data.table
+#' @useDynLib microbiomeComputations
+#' @export
+setGeneric("correlation",
+  function(data, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) standardGeneric("correlation"),
+  signature = c("data")
+)
+
+#'@export
+setMethod("correlation", signature("AbundanceData"), function(data, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+    df <- data@data
+    recordIdColumn <- data@recordIdColumn
+    naToZero <- data@imputeZero
+    ancestorIdColumns <- data@ancestorIdColumns
+    allIdColumns <- c(recordIdColumn, ancestorIdColumns)
+    sampleMetadata <- data@sampleMetadata
+
+    ## Initialize and check inputs
+    method <- veupathUtils::matchArg(method)
+    verbose <- veupathUtils::matchArg(verbose)
+
+    # Check that incoming df meets requirements
+    if (!'data.table' %in% class(df)) {
+      data.table::setDT(df)
+    }
+
+    # Check that incoming metadata meets requirements
+    if (!'data.table' %in% class(sampleMetadata)) {
+      data.table::setDT(sampleMetadata)
+    }
+
+    computeMessage <- ''
+    veupathUtils::logWithTime(paste("Received df table with", nrow(df), "samples and", (ncol(df)-1), "taxa."), verbose)
+
+    # Replace NA values with 0
+    if (naToZero) {
+      veupathUtils::setNaToZero(df)
+      veupathUtils::logWithTime("Replaced NAs with 0", verbose)
+    }
+
+    # Check that the order of sample ids match. If not, reorder so they do
+    if (!identical(df[[recordIdColumn]], sampleMetadata[[recordIdColumn]])) {
+      # Reorder sampleMetadata so that it matches the order in df
+      sampleMetadata <- sampleMetadata[match(df[[recordIdColumn]],sampleMetadata[[recordIdColumn]]), ]
+      veupathUtils::logWithTime("Reordered sampleMetadata rows based on abundance data sample id order.", verbose)
+    }
+
+    ## Find numeric metadata
+    numericMetadataCols <- veupathUtils::findNumericCols(sampleMetadata)
+    if (length(numericMetadataCols) < 1) {
+      stop("correlation requires at least one continuous metadata variable.")
+    }
+
+
+    ## Compute correlation
+    # Resulting data table has column "rn" = row names of the correlation matrix (so taxa names), and the other
+    # column names are the vars from sample metadata that we use
+    corrResult <- data.table::as.data.table(cor(df[, -..allIdColumns], sampleMetadata[, ..numericMetadataCols], method = method), keep.rownames = T)
+
+    veupathUtils::logWithTime(paste0('Completed method=',method,'. Formatting results.'), verbose)
+
+
+    ## Format results
+    meltedCorrResult <- melt(corrResult, id.vars=c('rn'))
+    statistics <- data.frame(
+      var1 = meltedCorrResult[['rn']],
+      var2 = meltedCorrResult[['variable']],
+      correlationCoef = meltedCorrResult[['value']]
```

Comment:
can we be more explicit than var1 and var2?

Reply:
i'd be happy to but i dont have a much better idea. The network uses terms like 'source' and 'target' but that sort of directionality isn't appropriate here. I also thought about 'correlate1' and 'correlate2' but that is hard to distinguish from the verb "correlate" lol and also seems like i'm already assuming these two variables are correlated. 
We could at least say "variable1" and "variable2" though?

Reply:
when we add support for metadata vs metadata and assay vs assay, do we see multiple correlation methods w different signatures? is it possible to explicitly say one of these are abundance data and the other sample metadata?

Reply:
that's a good question. Yes i expect different signatures so definitely possible. The other question i have though (thinking aloud here) is does having the abundance and metadata names here in the stats df help us or make some code further down the line harder because it has to account for lots of different stats col names? Like then the api would have to account for different col names. Totally possible, but then in the ds it'll call out these columns explicitly in order to do some network things and though it'd be easier to have it all say the same var1, var2, it'll be easier to read and probably to validate if they have real names. Hmmmm by the end of this thought train i was at the opposite opinion of where i started.

Reply:
fwiw i also dont like calling it var1 when its values are variables according to EDA verbiage..

Reply:
this approach will make it hard to validate there are only two variables and then the correlation results. So some stats object with assay, metadata, and another assay could get through. Also when we do correlation assay v assay we'll end up with the same problem and they'll be named assay1 and assay2. I could keep it variable1 and variable2 and then add another property that says var1 is an assay and var2 is metadata... hmmm